comms/src/dispatcher.rs

```diff
@@ -0,0 +1,188 @@
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use derive_error::Error;
+use std::{collections::HashMap, error::Error, hash::Hash};
+
+#[derive(Debug, Error, Clone)]
+pub enum DispatchError {
+    /// A dispatch route was not defined for the specific message type
+    MessageHandlerUndefined,
+    #[error(msg_embedded, non_std, no_from)]
+    HandlerError(String),
+}
+
+/// The signature of a handler function
+type HandlerFunc<M, E> = fn(msg: M) -> Result<(), E>;
+
+/// The trait bound for type parameter K on the dispatcher i.e the route key.
+/// K must be Eq + Hash + {able to be sent across threads} +
+/// 'static (all references must live as long as the program BUT there are no references
+/// so therefore this is simply to satisfy the closure on `thread::spawn`)
+/// This saves us from having to duplicate these trait bounds whenever we
+/// want specify the type parameter K (like a type alias).
+pub trait DispatchableKey: Eq + Hash + Send + 'static {}
+/// Implement this trait for all types which satisfy it's trait bounds.
+impl<T> DispatchableKey for T where T: Eq + Hash + Send + 'static {}
+
+/// A message type resolver. The resolver is called with the dispatched message.
+/// The resolver should then decide which dispatch key should be used.
+pub trait DispatchResolver<K, M>: Send + 'static
+where K: DispatchableKey
+{
+    fn resolve(&self, msg: &M) -> Result<K, DispatchError>;
+}
+
+/// Dispatcher pattern. Links handler function to "keys" which are resolved
+/// be a given type implementing [DispatchResolver].
+///
+/// ## Type Parameters
+/// `K` - The route key
+/// `M` - The type which is passed into the handler
+/// `E` - The type of error returned from the handler
+/// `R` - The resolver type
+#[derive(Clone, Debug)]
+pub struct Dispatcher<K, M, E, R>
+where K: DispatchableKey
+{
+    handlers: HashMap<K, HandlerFunc<M, E>>,
+    catch_all: Option<HandlerFunc<M, E>>,
+    resolver: R,
+}
+
+impl<K, M, E, R> Dispatcher<K, M, E, R>
+where
+    K: DispatchableKey,
+    R: DispatchResolver<K, M>,
+    E: Error,
+{
+    /// Construct a new MessageDispatcher with no defined dispatch routes
+    pub fn new(resolver: R) -> Dispatcher<K, M, E, R> {
+        Dispatcher {
+            handlers: HashMap::new(),
+            resolver,
+            catch_all: None,
+        }
+    }
+
+    /// This function allows a new dispatch route to be specified and added to the handlers, all received messaged that
+    /// are of the dispatch type will be routed to the specified handler_function
+    pub fn route(mut self, path_key: K, handler: HandlerFunc<M, E>) -> Self {
+        self.handlers.insert(path_key, handler);
+        self
+    }
+
+    /// Set the handler to use if no other handlers match
+    pub fn catch_all(mut self, handler: HandlerFunc<M, E>) -> Self {
```

Comment:
Setting a ```catch_all``` function is not needed as the catch all function can be set in the ```DispatchResolver``` as seen in the tests.

Reply:
I agree that this doesn't ever HAVE to be used, but it could be used to indicate what to do if there is no handler for a message type - Agreed you can put that in the resolver logic, but this could be neater - I'll leave it in for now (can't hurt)

Reply:
```rust
Dispatcher::new(...)
.route(M::A, do_a)
.route(M::B, do_b)
.route(M::DoEdgeCase, catch_all)
```

vs

```rust
Dispatcher::new(...)
.route(M::A, do_a)
.route(M::B, do_b)
.catch_all(catch_all)
```

You may not want an edge case "key" in your message types (see `TariMessage`)